dynamiqs/smesolve/rouchon.py

```diff
@@ -19,40 +19,32 @@ class SMERouchon1(SMERouchon):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
-        # define cached operators
-        # self.M0 (b_H, b_L, 1, 1, n, n)
-        self.M0_tmp = cache(lambda Hnh: self.I - 1j * self.dt * Hnh)
-
-        # define M1s
-        # self.M1s: # (len(L), 1, b_L, 1, 1, n, n)
+        self.M0_tmp = cache(lambda Hnh: self.I - 1j * self.dt * Hnh)  # (..., n, n)
         self.M1s = torch.cat(
             [
                 sqrt(self.dt) * self.Lc,
                 torch.sqrt(self.dt * (1 - self.etas))[..., None, None] * self.Lm,
             ],
-        )
+        )  # (nL, ..., n, n)
 
     def forward(self, t: float, rho: Tensor) -> Tensor:
-        # rho: (b_H, b_L, b_rho, ntrajs, n, n) -> (b_H, b_L, b_rho, ntrajs, n, n)
+        # rho: (..., n, n) -> (..., n, n)
 
-        # compute cached operators
-        # H, Hnh, M0_tmp: (b_H, b_L, 1, ntrajs, n, n)
-        H = self.H(t)
-        Hnh = self.Hnh(H)
-        M0_tmp = self.M0_tmp(Hnh)
+        H = self.H(t)  # (..., n, n)
+        Hnh = self.Hnh(H)  # (..., n, n)
+        M0_tmp = self.M0_tmp(Hnh)  # (..., n, n)
 
         # sample Wiener process
-        dw = self.sample_wiener(self.dt)  # (len(Lm), b_H, b_L, b_rho, ntrajs)
+        dw = self.sample_wiener(self.dt)  # (nLm, ...)
```

Comment:
the last dimension is still `ntrajs` no?

Reply:
Yes, but for smesolve I also included `ntrajs` into `...`. For cartesian batching we will have `... = bH, bL, by, ntrajs` and for stacked batching `... = bHLy, ntrajs`.